packages/sitecore-jss-nextjs/src/middleware/multisite-middleware.ts

```diff
@@ -0,0 +1,126 @@
+   * This is an important performance consideration since Next.js Edge middleware runs on every request.
+   * @param {string} pathname The pathname
+   * @returns {boolean} Whether to exclude the route
+   */
+  excludeRoute?: (pathname: string) => boolean;
+  /**
+   * function used to resolve site for given hostname
+   */
+  getSite: (hostname: string) => SiteInfo;
+  /**
+   * Fallback hostname in case `host` header is not present
+   * @default localhost
+   */
+  defaultHostname?: string;
+};
+
+/**
+ * Middleware / handler for multisite support
+ */
+export class MultisiteMiddleware {
+  private defaultHostname: string;
+
+  /**
+   * @param {MultisiteMiddlewareConfig} [config] Multisite middleware config
+   */
+  constructor(protected config: MultisiteMiddlewareConfig) {
+    this.defaultHostname = config.defaultHostname || 'localhost';
+  }
+
+  /**
+   * Gets the Next.js middleware handler with error handling
+   * @returns middleware handler
+   */
+  public getHandler(): (req: NextRequest, res?: NextResponse) => Promise<NextResponse> {
+    return async (req, res) => {
+      try {
+        return await this.handler(req, res);
+      } catch (error) {
+        console.log('Multisite middleware failed:');
+        console.log(error);
+        return res || NextResponse.next();
+      }
+    };
+  }
+
+  protected excludeRoute(pathname: string) {
+    if (
+      pathname.includes('.') || // Ignore files
+      pathname.startsWith('/api/') || // Ignore Next.js API calls
+      pathname.startsWith('/sitecore/') || // Ignore Sitecore API calls
+      pathname.startsWith('/_next') // Ignore next service calls
+    ) {
+      return true;
+    }
+    return false;
+  }
+
+  protected extractDebugHeaders(incomingHeaders: Headers) {
+    const headers = {} as { [key: string]: string };
+    incomingHeaders.forEach((value, key) => (headers[key] = value));
+    return headers;
+  }
+
+  private handler = async (req: NextRequest, res?: NextResponse): Promise<NextResponse> => {
+    const pathname = req.nextUrl.pathname;
+    const hostHeader = req.headers.get('host')?.split(':')[0];
+    const hostname = hostHeader || this.defaultHostname;
+
+    debug.multisite('multisite middleware start: %o', {
+      pathname,
+      hostname,
+    });
+
+    if (!hostHeader) {
+      debug.multisite(`host header is missing, default ${hostname} is used`);
+    }
+
+    // Response will be provided if other middleware is run before us
+    let response = res || NextResponse.next();
+
+    if (
+      this.excludeRoute(pathname) ||
+      (this.config.excludeRoute && this.config.excludeRoute(pathname))
+    ) {
+      debug.multisite('skipped (route excluded)');
+      return response;
+    }
+
+    // Site name can be forced by query string parameter
+    const siteName = req.nextUrl.searchParams.get('sc_site') || this.config.getSite(hostname).name;
```

Comment:
We should check for 'sc_site' cookie on request and use. If not, then do config.getSite, etc...